src/server/snapshot.cc

```diff
@@ -90,79 +90,40 @@ void SliceSnapshot::Start(bool stream_journal, const Cancellation* cll, Snapshot
-    if (cll->IsCancelled()) {
-      Cancel();
-    } else if (!stream_journal) {
-      CloseRecordChannel();
+    // We stop the channel if we are performing backups (non-streaming) or full sync failed.
+    // For a successful full-sync we keep the channel in order to switch to streaming mode.
+    if (cll->IsCancelled() || !stream_journal) {
+      StopChannel();
     }
   });
 }
 
 void SliceSnapshot::StartIncremental(Context* cntx, LSN start_lsn) {
-  auto* journal = db_slice_->shard_owner()->journal();
-  DCHECK(journal);
-
   serializer_ = std::make_unique<RdbSerializer>(compression_mode_);
 
-  snapshot_fb_ =
-      fb2::Fiber("incremental_snapshot", [this, journal, cntx, lsn = start_lsn]() mutable {
-        DCHECK(lsn <= journal->GetLsn()) << "The replica tried to sync from the future.";
-
-        VLOG(1) << "Starting incremental snapshot from lsn=" << lsn;
-
-        // The replica sends the LSN of the next entry is wants to receive.
-        while (!cntx->IsCancelled() && journal->IsLSNInBuffer(lsn)) {
-          serializer_->WriteJournalEntry(journal->GetEntry(lsn));
-          PushSerializedToChannel(false);
-          lsn++;
-        }
-
-        VLOG(1) << "Last LSN sent in incremental snapshot was " << (lsn - 1);
-
-        // This check is safe, but it is not trivially safe.
-        // We rely here on the fact that JournalSlice::AddLogRecord can
-        // only preempt while holding the callback lock.
-        // That guarantees that if we have processed the last LSN the callback
-        // will only be added after JournalSlice::AddLogRecord has finished
-        // iterating its callbacks and we won't process the record twice.
-        // We have to make sure we don't preempt ourselves before registering the callback!
-
-        // GetLsn() is always the next lsn that we expect to create.
-        if (journal->GetLsn() == lsn) {
-          {
-            FiberAtomicGuard fg;
-            serializer_->SendFullSyncCut();
-          }
-          auto journal_cb = absl::bind_front(&SliceSnapshot::OnJournalEntry, this);
-          journal_cb_id_ = journal->RegisterOnChange(std::move(journal_cb));
-          PushSerializedToChannel(true);
-        } else {
-          // We stopped but we didn't manage to send the whole stream.
-          cntx->ReportError(
-              std::make_error_code(errc::state_not_recoverable),
-              absl::StrCat("Partial sync was unsuccessful because entry #", lsn,
-                           " was dropped from the buffer. Current lsn=", journal->GetLsn()));
-          Cancel();
-        }
-      });
+  snapshot_fb_ = fb2::Fiber("incremental_snapshot", [cntx, start_lsn, this] {
+    this->SwitchIncrementalFb(cntx, start_lsn);
+  });
 }
 
-void SliceSnapshot::Stop() {
-  // Wait for serialization to finish in any case.
-  Join();
+void SliceSnapshot::Finalize() {
+  DCHECK(db_slice_->shard_owner()->IsMyThread());
+  DCHECK(journal_cb_id_);  // Called only for streaming use-case.
```

Comment:
why this is a DCHECK and if was removed? this is only for streaming so why this DCHECK does not fail in save?

Reply:
because we do not call this function in the `Save()` flow. this is the whole point of DCHECK - to help us understand how these functions are used. The code before was confusing because it "implied" that `journal_cb_id_` could be 0.

Reply:
Oh, I see my comment is wrong. I meant called only for replication usecases. I will fix the comments.